0005-ordered-quantification.md

```diff
@@ -0,0 +1,644 @@
+a sequence of Unicode code points that is not efficiently accessed with indices. The features proposed here provide ways to work with
+such datatypes efficiently at runtime.
+
+## Compiled let-such-that expressions without uniqueness proofs
+
+An expression `var x :| P(x); ...` is currently only allowed in compiled contexts
+if Dafny can prove that there is exactly one such `x` that satisfies `P(x)`.
+With ordered quantification in our toolbox, we could lift this requirement when
+the quantification is ordered, and ensure that the first such `x` according to that
+order is picked. This would be more useful if we also allowed the same additional syntax,
+i.e. `var x <- C :| P(x); ...`.
+
+## Array comprehensions
+
+The same syntax for sequence comprehensions could also be used for flexible array initialization,
+since this is another case where the user has to declare an ordered source of values:
+
+```
+var a := new int[10] i :: i * i;
+var a := new int[|s|] i :: s[i];     // Where s is a seq<int>
+var a := new int[] i | 10 <= i < 20;
+```
+
+It would likely be prudent to limit these cases to those where the size of the domain
+is easy to infer statically, so that the size of the array to allocate is known before
+enumerating the values.
+
+## Multiset comprehensions
+
+There would have similar semantics to sequence comprehensions, 
+where the multiplicity of results is significant but not the ordering.
+This feature would be very cheap to implement once sequence comprehensions are implemented.
+
+## Generalized comprehensions
+
+As shown above, sequence comprehensions are a powerfully expressive feature,
+but they can only express enumerating multiple values from a small number of source expressions, 
+and not the opposite direction of aggregating multiple values into one.
+The existing quantifier and comprehension expressions can be viewed as specific
+cases of aggregating multiple values into a single result value.
+An equally powerful feature would be to generalize this pattern and define
+a parameterized collection expression to aggregate a quantifier domain into
+a single value:
+
+```dafny
+collect(&&) x: T | P(x) :: Q(x)         == forall x: T | P(x) :: Q(x)
+collect(||) x: T | P(x) :: Q(x)         == exists x: T | P(x) :: Q(x)
+collect(SeqBuilder) x: T | P(x) :: Q(x) == seq x: T | P(x) :: Q(x)
+collect(SetBuilder) x: T | P(x) :: Q(x) == set x: T | P(x) :: Q(x)
+collect(+) x: T | P(x) :: Q(x)          == (summation)
+collect(*) x: T | P(x) :: Q(x)          == (product)
+collect(<) x: T | P(x) :: Q(x)          == (minimum)
+collect(Averager) x: T | P(x) :: Q(x)   == (average)
+collect(First(n)) x: T | P(x) :: Q(x)   == Take(seq x: T | P(x) :: Q(x), n)
+...
+```
+
+This mirrors the `Collector` concept from the Java streaming APIs, and ideally the shape of the
+parameter to `collect` expressions would define similar operations for merging intermediate results
+to leave the door open for parallel or even distributed computation.
```

Comment:
Would `collect` require a "zero" or "identity" element? Most such formalisms do, I think (such as `Monoid` in Haskell). In Dafny, we could avoid it by requiring `T` to be non-empty, though.

Reply:
I had that thought too. Formalizing the parameter to `collect` definitely needs a lot more thought and I didn't attempt to make this too concrete yet - it might work better to parameterize with a type parameter that implements a particular "type characteristic".